### PR TITLE
Feature/daef 251 Autocomplete input control

### PR DESCRIPTION
**Pull request by tomothespian**
This PR introduces autocomplete input control.

autocomplete input control

## TODOs
- [x] fix focus on label click
- [x] sort search results alphabetically (add property for this and use `true` as default)
- [x] allow duplicate entries (e.g. multiple selections of the same word)
- [x] limit number of visible results (add property for this and use `10` as default)
- [x] clear search after item selection (on mouse click in the suggestion list)
- [x] missing dropdown arrow
- [x] improve styling [Nikola]

### CODE REVIEW

**Comment by nikolaglumac** [comment]
Reviewing this now...

**Comment by DominikGuzei** [comment]
Me too 😉

**Comment by nikolaglumac** [comment]
@tomothespian new component works great!
I have just fixed some minor styling issues and also added few todo-s which I think should be fixed ;)

**Comment by DominikGuzei** [comment]
@tomothespian @nikolaglumac what about the open TODOs in the description?

**Comment by DominikGuzei** [comment]
thanks @nikolaglumac 👍

File: source/components/Autocomplete.js [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+  static propTypes = Object.assign({}, FormField.propTypes, {
+    error: PropTypes.string,
+    maxSelections: PropTypes.number,
+    placeholder: PropTypes.string,
+    suggestedWords: PropTypes.array,
+  });
+
+  state = {
+    selectedWords: [],
+    invalidCharacters: /[^a-zA-Z0-9]/g, // only allow letters and numbers
```
Comment (DominikGuzei) [inline source/components/Autocomplete.js]: I would either move this into a `const INVALID_CHARS` or make it configurable via a prop

File: source/components/Autocomplete.js [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+    } else {
+      this.closeSuggestions();
+    }
+  };
+
+  _handleDocumentClick = (event) => {
+    const root = this._getRootSkinPart();
+    const isDescendant = events.targetIsDescendant(event, ReactDOM.findDOMNode(root));
+    if (this.state.isSuggestionsOpened && !isDescendant) {
+      this.setState({ isSuggestionsOpened: false, highlightedOptionIndex: 0 });
```
Comment (DominikGuzei) [inline source/components/Autocomplete.js]: Call `this.closeSuggestions()` here

File: source/components/Autocomplete.js [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+      if (_.startsWith(suggestedWord, value)) {
+        filteredWords.push(suggestedWord);
+      }
+    });
+
+    this.setState({ filteredWords: (value !== '') ? filteredWords : this.props.suggestedWords });
+  };
+
+  updateSelectedWords = (event) => {
+    if (!this.props.maxSelections || (this.state.selectedWords.length < this.props.maxSelections && this.state.filteredWords && this.state.filteredWords.length > 0)) {
```
Comment (DominikGuzei) [inline source/components/Autocomplete.js]: Please shorten this condition with separate boolean vars like

`const canMoreWordsBeSelected = this.state.selectedWords.length < this.props.maxSelections`

File: source/skins/simple/AutocompleteSkin.js [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+          >
+            &times;
+          </span>
+        </span>
+      </span>
+    );
+  });
+
+  // show placeholder only if no maximum selections declared or maximum not reached
+  let placeholder = (!props.maxSelections || (props.selectedWords.length < props.maxSelections)) ? props.placeholder : '';
```
Comment (DominikGuzei) [inline source/skins/simple/AutocompleteSkin.js]: you can use the same boolean here to make the code more readable 😉

**Comment by tomothespian** [comment]
No problem guys, will improve code following your suggestions and TODO's.  ;)

**Comment by tomothespian** [comment]
@DominikGuzei  fixes and improvements are done ;)

**Comment by darko-mijic** [comment]
@tomothespian `INVALID_CHARS` a constant now but I think we also need to receive this via prop. This is proably the first thing that will be customised. A default value can be the value of the constant. 
Can we also update the readme with this component? And others which are missing there.

File: source/components/Autocomplete.js [hunk 5]
```diff
@@ -139,15 +152,19 @@ export default class Autocomplete extends FormField {
   };
 
   updateSelectedWords = (event) => {
-    if (!this.props.maxSelections || (this.state.selectedWords.length < this.props.maxSelections && this.state.filteredWords && this.state.filteredWords.length > 0)) {
+    const canMoreWordsBeSelected = this.state.selectedWords.length < this.props.maxSelections;
+
+    if (!this.props.maxSelections || (canMoreWordsBeSelected && this.state.filteredWords && this.state.filteredWords.length > 0)) {
```
Comment (DominikGuzei) [inline source/components/Autocomplete.js]: You can shorten the second boolean too 😉 
`const areFilteredWordsAvailable = this.state.filteredWords && this.state.filteredWords.length > 0;`

**Comment by DominikGuzei** [comment]
Reviewing again now …

**Comment by DominikGuzei** [comment]
Great work @tomothespian, all mentioned problems have been fixed and i will merge now!

**Pull request merged**